ocrd_tesserocr/common.py

```diff
@@ -5,43 +5,145 @@
 import io
 
 import numpy as np
-from PIL import Image, ImageDraw
+from PIL import Image, ImageDraw, ImageStat
 
 from ocrd_utils import getLogger, xywh_from_points, polygon_from_points
 
 LOG = getLogger('') # to be refined by importer
 
 # to be refactored into core (as function in ocrd_utils):
 def polygon_mask(image, coordinates):
+    """"Create a mask image of a polygon.
+    
+    Given a PIL.Image `image` (merely for dimensions), and
+    a numpy array `polygon` of relative coordinates into the image,
+    create a new image of the same size with black background, and
+    fill everything inside the polygon hull with white.
+    
+    Return the new PIL.Image.
+    """
     mask = Image.new('L', image.size, 0)
+    if isinstance(coordinates, np.ndarray):
+        coordinates = list(map(tuple, coordinates))
     ImageDraw.Draw(mask).polygon(coordinates, outline=1, fill=255)
     return mask
 
 # to be refactored into core (as function in ocrd_utils):
-def rotate_polygon(coordinates, angle, orig={'x': 0, 'y': 0}):
-    # if the region image has been rotated, we must also
-    # rotate the coordinates of the line
-    # (which relate to the top page image)
-    # in the same direction but with inverse transformation
-    # matrix (i.e. passive rotation), and
-    # (since the region was rotated around its center,
-    #  but our coordinates are now relative to the top left)
-    # by first translating to center of region, then
-    # rotating around that center, and translating back:
-    # point := (point - region_center) * region_rotation + region_center
-    # moreover, since rotation has reshaped/expanded the image,
-    # the line coordinates must be offset by those additional pixels:
-    # point := point + 0.5 * (new_region_size - old_region_size)
+def image_from_polygon(image, polygon):
+    """"Mask an image with a polygon.
+    
+    Given a PIL.Image `image` and a numpy array `polygon`
+    of relative coordinates into the image, fill everything
+    outside the polygon hull to the background. Since `image`
+    is not necessarily binarized yet, determine the background
+    from the median color (instead of white).
+    
+    Return a new PIL.Image.
+    """
+    mask = polygon_mask(image, polygon)
+    # create a background image from its median color
+    # (in case it has not been binarized yet):
+    # array = np.asarray(image)
+    # background = np.median(array, axis=[0, 1], keepdims=True)
+    # array = np.broadcast_to(background.astype(np.uint8), array.shape)
+    background = ImageStat.Stat(image).median[0]
```

Comment:
This could turn nasty in edge cases, right? Do we have use cases where non-binarized images are masked?

Reply:
see below under discussion of transparency instead of median